### PR TITLE
Maja Octos Grocery_Store

### PR DESCRIPTION
**Pull request by mgraonic**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Question 	|   Response	|
|---	|---	|
|   Why is it useful to put classes inside modules? 	|   To group together related classes and methods without running into naming collisions.	|
|   What is accomplished with `raise ArgumentError`? 	|   We were told not to do this for the grocery-store exercise. Generally an argument error is raised when attempting to pass an incorrect number of arguments or pass an incorrect argument data type.	|
|   Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods?	|  Needed to access a piece of information (attribute) across all classes, not just from a single instance. 	|
|   Why does it make sense to use inheritance for the online order?  | The online order will have a lot of the same behavior so it's not necessary to retype that code. I'm sure there are some efficiency considerations as well.  |
|   Did the presence of automated tests change the way you thought about the problem? How? | Yes, but I felt very clueless about how to write the test (when/how to instantiate an object or objects to be tested, how to write meaningful tests that account for nominal and edge cases, etc. |

### CODE REVIEW

**Comment by mgraonic** [comment]
I wasn't sure how the self.all and self.find methods were supposed to interact. Should the self.find method traverse the CSV file itself instead of the array created by self.all?

File: lib/order.rb [hunk 1]
```diff
@@ -1,18 +1,74 @@
...
     end
 
+    def self.find(id)
+      # finds an order in the collection of Order instances
+      @@orders.each do |order|
+        if order.id == id
+          return order
+        end
+      end
+      return "ERROR: order does not exist"
```
Comment (droberts-sea) [inline lib/order.rb:24]: Is it really an error if the order doesn't exist? A more friendly strategy might be to return `nil`, and let whoever called the method check for that value.

File: lib/order.rb [hunk 2]
```diff
@@ -1,18 +1,74 @@
...
     def initialize(id, products)
-      @id = id
+      @id = id.to_i
       @products = products
     end
 
+    def self.find(id)
+      # finds an order in the collection of Order instances
+      @@orders.each do |order|
+        if order.id == id
```
Comment (droberts-sea) [inline lib/order.rb:20]: Referencing `@@orders` here means that `.all` must be called before `.find`. This is a dependency that we can avoid!

Instead of saying `@@orders.each do`, you could say `self.all.each do`. This will call the `all` method, which will read the orders from the file if that hasn't happened yet.

This is an example of what Metz is talking about in chapter 2 when she says "depend on behavior, not data".

File: lib/order.rb [hunk 3]
```diff
@@ -1,18 +1,74 @@
...
+
+            id = line_item[0]
+            items_string = line_item[1]
+            products = {} #will take k/v pairs delimited by colon
+
+            semicolon_split = items_string.split(';')
+
+            semicolon_split.each do |string|
+              key_value_split = string.split(':')
+              products[key_value_split[0]] = key_value_split[1].to_f
```
Comment (droberts-sea) [inline lib/order.rb:43]: Good work getting this tricky logic sorted out.

File: lib/order.rb [hunk 4]
```diff
@@ -1,18 +1,74 @@
+require 'csv'
+require 'awesome_print'
+
 module Grocery
   class Order
     attr_reader :id, :products
 
+    FILE_NAME = "support/orders.csv"
+
```
Comment (droberts-sea) [inline lib/order.rb:9]: Good use of a constant here!

File: specs/order_spec.rb [hunk 5]
```diff
@@ -78,33 +79,53 @@
   end
 end
 
-# TODO: change 'xdescribe' to 'describe' to run these tests
-xdescribe "Order Wave 2" do
+describe "Order Wave 2" do
+  before do
+    @all_orders = Grocery::Order.all
+    # @orders = Grocery::Order
```
Comment (droberts-sea) [inline specs/order_spec.rb:85]: Calling `Order.all` here masks that dependency between `.find` and `.all` that I discussed above.

File: specs/online_order_spec.rb [hunk 6]
```diff
@@ -1,90 +1,112 @@
...
-      # Instatiate your OnlineOrder here
-      # online_order =
-      # online_order.must_be_kind_of Grocery::Order
-    end
-
-    it "Can access Customer object" do
-      # TODO: Your test code here!
+      @all_online_orders.each do |order|
+        order.must_be_kind_of Grocery::Order
+      end
```
Comment (droberts-sea) [inline specs/online_order_spec.rb:17]: Shouldn't this be `order.must_be_kind_of Grocery::OnlineOrder`?

The reason this test passes is because an `OnlineOrder` is a kind of `Order`.

File: lib/online_order.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  class OnlineOrder < Grocery::Order
+    attr_reader :id, :products, :customer_id, :status, :total, :add_product, :find_by_customer
+
+    FILE_NAME = "support/online_orders.csv"
+
+    @@online_orders = []
+
+    def initialize(id, products, customer_id, status = :pending)
+      super(id, products)
+      @customer_id = customer_id
```
Comment (droberts-sea) [inline lib/online_order.rb:15]: Good use of `super` here

File: lib/online_order.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        return super() + 10
+      end
+    end
+
+    def add_product(product_name, product_price)
+      unless [:pending, :paid].include?(status)
+        return nil
+      else
+        @products[product_name] = product_price
+      end
```
Comment (droberts-sea) [inline lib/online_order.rb:78]: This misses the behavior we had with `Order`, where if the product was already in the order the method would return false. One way to get that behavior would be to call `super` here.

File: lib/online_order.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    end
+
+    def self.find_by_customer(customer_id)
+      customer_orders = []
+      @@online_orders.each do |order|
+        if order.customer_id == customer_id
+          customer_orders << order
+        end
+      end
+      return customer_orders
```
Comment (droberts-sea) [inline lib/online_order.rb:35]: I like that this will return an empty array if no orders match.

**Comment by droberts-sea** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** | 
Answered comprehension questions | yes
Used Git Regularly | yes
**Wave 1** | 
All provided tests pass | yes
Using the appropriate `attr_` for instance variables | yes
**Wave 2** | 
All stubbed tests are implemented fully and pass | yes
Appropriately parses the product data from CSV file in `Order.all` | yes
Used `CSV` library only in `Order.all` (not in `Order.find`) | yes
Used `Order.all` to get order list in `Order.find` | no - see inline
**Wave 3** | 
All stubbed tests are implemented fully and pass | yes
Used inheritance in the initialize for online order | yes
Used inheritance for the total method in online order | yes
Use `CSV` library only in `OnlineOrder.all` | yes
Used `all` to get order list in `find` | no
Appropriately searches for Customer orders in `find_by_customer` | yes
**Additional Notes** |

Great job overall! This code is clean and well-organized. I've left some inline comments below for you to review, but in general I am happy wth this submission.

Regarding tests, your tests on this project look good to me, but if you want a little more guidance, remember that _Arrange, Act, Assert_ pattern we talked about in class.
1. _Arrange_ things to your liking
    - Set initial values
    - Create an instance of the class you're testing
    - Calculate expected results
    - Do any other setup
1. _Act_
    - Call the method you're currently testing
1. _Assert_
    - Write expectations to ensure things came out the way you wanted

Remember also that you don't need to nail all your test cases the first time - every time you fix a bug or stumble over some bad behavior is an opportunity to write another test. This is certainly a skill that takes practice to get comfortable with. Keep up the hard work!
